org/mozilla/jss/nss/SSL.c

```diff
@@ -0,0 +1,215 @@
+#include <nspr.h>
+#include <nss.h>
+#include <ssl.h>
+#include <limits.h>
+#include <stdint.h>
+#include <jni.h>
+
+#include "java_ids.h"
+#include "jssutil.h"
+#include "pk11util.h"
+#include "PRFDProxy.h"
+
+#include "_jni/org_mozilla_jss_nss_SSL.h"
+
+jobject JSS_NewSecurityStatusResult(JNIEnv *env, int on, char *cipher,
+    int keySize, int secretKeySize, char *issuer, char *subject)
+{
+    jclass resultClass;
+    jmethodID constructor;
+    jobject result = NULL;
+    jbyteArray cipher_java = NULL;
+    jbyteArray issuer_java = NULL;
+    jbyteArray subject_java = NULL;
+
+    PR_ASSERT(env != NULL);
+
+    resultClass = (*env)->FindClass(env, "org/mozilla/jss/nss/SecurityStatusResult");
+    if (resultClass == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    constructor = (*env)->GetMethodID(env, resultClass, PLAIN_CONSTRUCTOR,
+        "(I[BII[B[B)V");
+    if (constructor == NULL) {
+        ASSERT_OUTOFMEM(env);
+        goto finish;
+    }
+
+    if (cipher) {
+        cipher_java = JSS_ToByteArray(env, cipher, strlen(cipher));
+    }
+
+    if (issuer) {
+        issuer_java = JSS_ToByteArray(env, issuer, strlen(issuer));
+    }
+
+    if (subject) {
+        subject_java = JSS_ToByteArray(env, subject, strlen(subject));
+    }
+
+    result = (*env)->NewObject(env, resultClass, constructor, on, cipher_java,
+        keySize, secretKeySize, issuer_java, subject_java);
+
+finish:
+    return result;
+}
+
+JNIEXPORT jobject JNICALL
+Java_org_mozilla_jss_nss_SSL_ImportFD(JNIEnv *env, jclass clazz, jobject model,
+    jobject fd)
+{
+    PRFileDesc *result = NULL;
+    PRFileDesc *real_model = NULL;
+    PRFileDesc *real_fd = NULL;
+
+    PR_ASSERT(env != NULL);
+
+    if (model != NULL && JSS_PR_getPRFileDesc(env, model, &real_model) != PR_SUCCESS) {
+        return NULL;
+    }
+
+    if (fd != NULL && JSS_PR_getPRFileDesc(env, fd, &real_fd) != PR_SUCCESS) {
+        return NULL;
+    }
+
+    result = SSL_ImportFD(real_model, real_fd);
+
+    return JSS_PR_wrapPRFDProxy(env, &result);
+}
+
+JNIEXPORT int JNICALL
+Java_org_mozilla_jss_nss_SSL_OptionSet(JNIEnv *env, jclass clazz, jobject fd,
+    jint option, jint val)
+{
+    PRFileDesc *real_fd = NULL;
+
+    PR_ASSERT(env != NULL && fd != NULL);
```

Comment:
We should really do better error handling on `fd == NULL` and raise a proper Java exception.